### PR TITLE
feat: connect to already deployed contracts

### LINKED ISSUE #88: Connect contract instance to existing contract
Currently, we're only able to create a contract instance by compiling and deploying it: 

```rust
let contract_instance = MyContract::new(compiled, client);
```

We should be able to create a contract instance from an already deployed contract by referencing to its deployed address, similar to `ethers.js` / `ethers-rs`. Something similar to this:

```rust
 let contract_instance = MyContract::connect("0x...");
```

### PR DESCRIPTION
**Pull request by vnepveu**
Closes #88.
This PR changes the way to instantiate a contract. Since the compiled code is actually not needed
to interact with the contract, we just provide the `contract_id`. 
This PR also creates a "default instantiation" for contracts, mainly for test purposes.

Comment: not sure about some design choices around string or `new_default`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -163,7 +163,7 @@ impl MyContract {
     pub fn initialize_counter(&self, arg: u64) -> ContractCall<u64> {
         Contract::method_hash(
             &self.fuel_client,
-            &self.compiled,
+            &self.contrat_id,
```
Comment (nfurfaro) [inline README.md:166]: typo : contract_id

File: README.md [hunk 2]
```diff
@@ -173,7 +173,7 @@ impl MyContract {
     pub fn increment_counter(&self, arg: u64) -> ContractCall<u64> {
         Contract::method_hash(
             &self.fuel_client,
-            &self.compiled,
+            &self.contrat_id,
```
Comment (nfurfaro) [inline README.md:176]: same typo as above.

**Review by adlerjohn — COMMENTED** [review]
Looks reasonable. Will defer to @digorithm for final review.

File: fuels-core/src/code_gen/abigen.rs [hunk 3]
```diff
@@ -111,18 +111,25 @@ impl Abigen {
...
                     impl #name {
-                        pub fn new(compiled: CompiledContract, fuel_client: FuelClient) -> Self {
-                            Self{ compiled, fuel_client }
+                        pub fn new(contract_id: String, fuel_client: FuelClient) -> Self {
+                            let contract_id = ContractId::from_str(&contract_id).unwrap();
+                            Self{ contract_id, fuel_client }
+                        }
+                        pub fn new_default(fuel_client: FuelClient) -> Self {
+                            let contract_id = ContractId::new([0xFFu8;32]);
+                            Self{ contract_id, fuel_client }
```
Comment (digorithm) [inline fuels-core/src/code_gen/abigen.rs:132]: Can we document this? It's not immediately clear, from the name alone, what the differences are between instantiating a contract with `new` and `new_default`.

**Review by digorithm — CHANGES_REQUESTED** [review]
Looks good, great job removing the need for passing the compiled to contract around, much better with just the contract id -- just left a tiny request for comment.

File: fuels-core/src/code_gen/abigen.rs [hunk 4]
```diff
@@ -111,18 +111,25 @@ impl Abigen {
...
 
                     impl #name {
-                        pub fn new(compiled: CompiledContract, fuel_client: FuelClient) -> Self {
-                            Self{ compiled, fuel_client }
+                        pub fn new(contract_id: String, fuel_client: FuelClient) -> Self {
+                            let contract_id = ContractId::from_str(&contract_id).unwrap();
+                            Self{ contract_id, fuel_client }
+                        }
+                        pub fn new_default(fuel_client: FuelClient) -> Self {
+                            let contract_id = ContractId::new([0xFFu8;32]);
```
Comment (adlerjohn) [inline fuels-core/src/code_gen/abigen.rs:131]: What is the reason for this function?

Comment (digorithm) [reply fuels-core/src/code_gen/abigen.rs:131]: I'm asking myself if there's a real good reason for it to exist. The only reason it exists is that in our test suite (`harness.rs`) we have some tests that instantiate a contract but don't make actual calls -- they're only testing stuff like the encoded data for a generated method call. So we're not deploying it. I'm guessing this `new_default` covers this need _but_ I guess we could get away by just having `new()` that takes the contract id, and in the tests that are not deploying, we just pass a zeroed contract id, like `new_default()` is doing; simplifying the whole API.

Comment (adlerjohn) [reply fuels-core/src/code_gen/abigen.rs:131]: Could `new_default` be non-`pub` if it's only to be used for testing? Or can the casts `impl` a `new_default` only in the test suite for this struct?

Comment (vnepveu) [reply fuels-core/src/code_gen/abigen.rs]: The idea of this `new_default` is indeed to allow "cleaner" testing. Upon reading your comments, I realize it's not good to have this function exist in *every* contract. I pushed the alternative, which is to feed a *null* `contract_id` to the tests.

File: Cargo.lock [hunk 5]
```diff
@@ -1,4444 +0,0 @@
-# This file is automatically @generated by Cargo.
```
[the remark this answers is not in the document]
Comment (vnepveu) [reply Cargo.lock:1]: Uh I messed up I think

File: fuels-abigen-macro/Cargo.toml [hunk 6]
```diff
@@ -26,4 +26,4 @@ fuels-contract = { version = "0.4.3", path = "../fuels-contract" }
 fuels-signers = { version = "0.4.3", path = "../fuels-signers", features=["test-helpers"] }
 hex = { version = "0.4.3", default-features = false, features = ["std"] }
 sha2 = "0.9.5"
-tokio = "1.15.0"
+tokio = "1.15.0"
```
Comment (adlerjohn) [inline fuels-abigen-macro/Cargo.toml:29]: newline

If you're using vscode, turn on adding newline at end of file

Comment (vnepveu) [reply fuels-abigen-macro/Cargo.toml:29]: It's the rustfmt formatter it's weird

**Review by digorithm — APPROVED** [review]
LGTM. Let's keep in mind that this is a breaking change -- we're moving from creating a contract instance with the compiled contract to taking only the contract id. All tests out there will break with this change. It will require minimal changes on the user side, though, similar to the changes in `harness.rs` in this PR.

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**